File: Sources/Server/Sources/Game/Entity/Dynamics/Mob/MobAggressivePursuit.ts
import { MobType } from "../../../../../../Shared/Entity/Statics/EntityType";
import { WavePool } from "../../../Genres/Wave/WavePool";
import { isPetal, isBody, isDeadEntity } from "../../../Utils/common";
import { Entity, EntityMixinConstructor, EntityMixinTemplate, onUpdateTick } from "../Entity";
import { BaseMob } from "./Mob";

export function findNearestEntity<T extends Entity>(me: T, entities: T[]): T | null {
    if (!entities.length) return null;

    return entities.reduce((nearest, current) => {
        const distanceToCurrent = Math.hypot(
            current.x - me.x,
            current.y - me.y,
        );

        const distanceToNearest = Math.hypot(
            nearest.x - me.x,
            nearest.y - me.y,
        );

        return distanceToCurrent < distanceToNearest ? current : nearest;
    });
}

export function turnAngleToTarget(thisAngle: number, dx: number, dy: number): number {
    const targetAngle = (Math.atan2(dy, dx) * 40.549) % 255; // 255/(2*PI)≈40.549
    const normalizedAngle = ((thisAngle % 255) + 255) % 255;
    let angleDiff = targetAngle - normalizedAngle;

    if (angleDiff > 127.5) angleDiff -= 255;
    else if (angleDiff < -127.5) angleDiff += 255;

    return ((normalizedAngle + angleDiff * 0.1 + 255) % 255);
}

const MOB_DETECTION_RANGE = 25;

export enum MobBehavior {
    Aggressive,
    Passive,
    Cautions,
    Neutral,
    None,
}

export const MOB_BEHAVIORS = {
    [MobType.Starfish]: MobBehavior.Aggressive,
    [MobType.Beetle]: MobBehavior.Aggressive,
    [MobType.Bubble]: MobBehavior.Passive,
    [MobType.Jellyfish]: MobBehavior.Cautions,
    [MobType.Bee]: MobBehavior.Neutral,
    [MobType.Spider]: MobBehavior.Aggressive,

    [MobType.Centipede]: MobBehavior.None,
    // TODO: elucidate desert centipede move
    [MobType.CentipedeDesert]: MobBehavior.Neutral,
    [MobType.CentipedeEvil]: MobBehavior.Aggressive,
} satisfies Record<MobType, MobBehavior>;

export function MobAggressivePursuit<T extends EntityMixinConstructor<BaseMob>>(Base: T) {
    return class extends Base implements EntityMixinTemplate {
        [onUpdateTick](poolThis: WavePool): void {
            super[onUpdateTick](poolThis);

            // Dont chase player when this is petal
            if (isPetal(this.type)) return;

            // If body, dont do anything
            if (isBody(poolThis, this)) return;

            // Dont do anything while this.starfishRegeningHealth so
            // can handle angle in MobHealthRegen.ts
            if (this.starfishRegeningHealth) return;

            // Target entity dead, stop target
            if (this.targetEntity && isDeadEntity(poolThis, this.targetEntity)) this.targetEntity = null;

            // Last attacked entity dead, stop target
            if (this.lastAttackedEntity && isDeadEntity(poolThis, this.lastAttackedEntity)) this.lastAttackedEntity = null;

            let distanceToTarget = 0;
            if (this.targetEntity) {
                const dx = this.targetEntity.x - this.x;
                const dy = this.targetEntity.y - this.y;
                distanceToTarget = Math.hypot(dx, dy);
            }

            // Lose entity
            if (distanceToTarget > this.loseRange) this.targetEntity = null;

            // Select target
            let targets: Entity[];

            if (this.petMaster) {
                // Mob which summoned by player will attack other mobs expect me, petals and pets
                targets = poolThis.getAllMobs().filter(p =>
                    !(
                        p.id === this.id ||
                        isPetal(p.type) ||
                        p.petMaster
                    ),
                );
            } else {
                // Target living players, p̶e̶t̶s̶
                // Mob will never target pets before wave
                targets = poolThis.getAllClients().filter(p => !p.isDead);
            }

            switch (MOB_BEHAVIORS[this.type]) {
                // Aggressive
                case MobBehavior.Aggressive: {
                    const targetableEntity = this.targetEntity || findNearestEntity(this, targets);
                    if (targetableEntity) {
                        const dx = targetableEntity.x - this.x;
                        const dy = targetableEntity.y - this.y;
                        const distance = Math.hypot(dx, dy);

                        if (distance < this.detectionRange) {
                            this.angle = turnAngleToTarget(
                                this.angle,
                                dx,
                                dy,
                            );

                            this.magnitude = 255 * this.speed;

                            this.targetEntity = targetableEntity;
                        } else {
                            this.targetEntity = null;
                        }
                    } else {
                        this.targetEntity = null;
                    }

                    break;
                }

                // Cautious (jellyfish)
                case MobBehavior.Cautions: {
                    const targetableEntity = this.targetEntity || findNearestEntity(this, targets);
                    if (targetableEntity) {
                        const dx = targetableEntity.x - this.x;
                        const dy = targetableEntity.y - this.y;
                        const distance = Math.hypot(dx, dy);

                        if (distance < this.detectionRange) {
                            this.angle = turnAngleToTarget(
                                this.angle,
                                dx,
                                dy,
                            );

                            this.magnitude = 255 * (this.targetEntity && distanceToTarget < (3 * this.size) ? 0 : 2);

                            this.targetEntity = targetableEntity;
                        } else {
                            this.targetEntity = null;
                        }
                    } else {
                        this.targetEntity = null;
                    }

                    break;
                }

                // Passive (bubble, stone)
                case MobBehavior.Passive: {
                    this.magnitude = 0;

                    break;
                }

                // Neutral (bee)
                case MobBehavior.Neutral: {
                    if (this.lastAttackedEntity) {
                        const dx = this.lastAttackedEntity.x - this.x;
                        const dy = this.lastAttackedEntity.y - this.y;

                        this.angle = turnAngleToTarget(
                            this.angle,
                            dx,
                            dy,
                        );

                        this.magnitude = 255 * this.speed;

                        this.targetEntity = this.lastAttackedEntity;
                    } else {
                        this.targetEntity = null;
                    }

                    break;
                }

                // Do nothing
                case MobBehavior.None: break;
            }
        }

        /**
         * Get detection range within mob.
         */
        get detectionRange(): number {
            return MOB_DETECTION_RANGE * this.size;
        }

        /**
         * Get lose range within mob.
         */
        get loseRange(): number {
            return (MOB_DETECTION_RANGE * 2) * this.size;
        }

        dispose(): void {
            if (super.dispose) {
                super.dispose();
            }
        }
    };
}

File: Sources/goserver/internal/wave/mob_aggressive_pursuit.go
package wave

import (
	"math"
	"math/rand/v2"
	"time"

	"flooooio/internal/native"
)

const (
	jellyfishLightningShootMS = 1000
)

// FindNearestEntity finds the nearest entity from a slice of entities.
func FindNearestEntity(me Node, entities []Node) Node {
	if len(entities) == 0 {
		return nil
	}

	// TODO: this should be nil
	nearest := entities[0]
	for _, current := range entities[1:] {
		distanceToCurrent := math.Hypot(
			current.GetX()-me.GetX(),
			current.GetY()-me.GetY(),
		)

		distanceToNearest := math.Hypot(
			nearest.GetX()-me.GetX(),
			nearest.GetY()-me.GetY(),
		)

		if distanceToCurrent < distanceToNearest {
			nearest = current
		}
	}

	return nearest
}

func FindNearestEntityWithLimitedDistance(me Node, entities []Node, maxDistance float64) Node {
	if len(entities) == 0 {
		return nil
	}

	var nearest Node
	nearestDistance := maxDistance

	for _, current := range entities {
		distanceToCurrent := math.Hypot(
			current.GetX()-me.GetX(),
			current.GetY()-me.GetY(),
		)

		if distanceToCurrent <= maxDistance && (nearest == nil || distanceToCurrent < nearestDistance) {
			nearest = current
			nearestDistance = distanceToCurrent
		}
	}

	return nearest
}

const angleFactor = 40.5845104884 // 255/(2*PI)

// TurnAngleToTarget calculates interpolated angle to target.
func TurnAngleToTarget(thisAngle, dx, dy float64) float64 {
	targetAngle := math.Mod(math.Atan2(dy, dx)*angleFactor, 255)
	normalizedAngle := math.Mod(math.Mod(thisAngle, 255)+255, 255)
	angleDiff := targetAngle - normalizedAngle

	if angleDiff > 127.5 {
		angleDiff -= 255
	} else if angleDiff < -127.5 {
		angleDiff += 255
	}

	return math.Mod(normalizedAngle+angleDiff*0.1+255, 255)
}

const mobDetectionRange = 15.

func getDetectionRange(m *Mob) float64 {
	return mobDetectionRange * m.DesiredSize()
}

func getLoseRange(m *Mob) float64 {
	return (mobDetectionRange * 2) * m.DesiredSize()
}

func getTargetNodes(wp *WavePool, m *Mob) []Node {
	var targets []Node

	if m.PetMaster != nil {
		mobs := wp.GetMobsWithCondition(func(fm *Mob) bool {
			return fm.Id != m.Id && fm.PetMaster == nil
		})

		targets = make([]Node, len(mobs))
		for i, mob := range mobs {
			targets[i] = mob
		}
	} else {
		players := wp.GetPlayersWithCondition(func(p *Player) bool {
			return !p.IsDead
		})

		targets = make([]Node, len(players))
		for i, player := range players {
			targets[i] = player
		}
	}

	return targets
}

func (m *Mob) MobAggressivePursuit(wp *WavePool) {
	// If body, dont do anything
	if IsBody(wp, m) {
		return
	}

	// Dont do anything while p.StarfishRegeningHealth so
	// can handle angle in mob_health_regen.go
	if m.StarfishRegeningHealth {
		return
	}

	// Target entity dead, stop target
	if m.TargetEntity != nil && IsDeadNode(wp, m.TargetEntity) {
		m.TargetEntity = nil
	}

	// Last attacked entity dead, stop target
	if m.LastAttackedEntity != nil && IsDeadNode(wp, m.LastAttackedEntity) {
		m.LastAttackedEntity = nil
	}

	distanceToTarget := 0.
	if m.TargetEntity != nil {
		dx := m.TargetEntity.GetX() - m.X
		dy := m.TargetEntity.GetY() - m.Y

		distanceToTarget = math.Hypot(dx, dy)
	}

	// Lose entity
	if distanceToTarget > getLoseRange(m) {
		m.TargetEntity = nil
	}

	switch native.EachMobBehaviorDefinition[m.Type] {
	case native.NoneBehavior:
		return

	case native.AggressiveBehavior:
		{
			var targetEntity Node
			if m.TargetEntity != nil {
				targetEntity = m.TargetEntity
			} else {
				targetEntity = FindNearestEntity(m, getTargetNodes(wp, m))
			}

			if targetEntity == nil {
				return
			}

			dx := targetEntity.GetX() - m.X
			dy := targetEntity.GetY() - m.Y
			distance := math.Hypot(dx, dy)

			if getDetectionRange(m) > distance {
				m.Angle = TurnAngleToTarget(
					m.Angle,
					dx,
					dy,
				)

				m.Magnitude = m.Speed() * 255

				m.TargetEntity = targetEntity
			} else {
				m.TargetEntity = nil
			}
		}

	case native.CautionBehavior:
		{
			var targetEntity Node
			if m.TargetEntity != nil {
				targetEntity = m.TargetEntity
			} else {
				targetEntity = FindNearestEntity(m, getTargetNodes(wp, m))
			}

			if targetEntity == nil {
				return
			}

			dx := targetEntity.GetX() - m.X
			dy := targetEntity.GetY() - m.Y
			distance := math.Hypot(dx, dy)

			if getDetectionRange(m) > distance {
				m.Angle = TurnAngleToTarget(
					m.Angle,
					dx,
					dy,
				)

				shouldStop := m.TargetEntity != nil && (3*m.Size) > distanceToTarget

				m.TargetEntity = targetEntity

				var magnitude float64
				if shouldStop {
					magnitude = 0
				} else {
					magnitude = m.Speed() * 255
				}

				m.Magnitude = magnitude

				switch v := targetEntity.(type) {
				case *Player:
					{
						// Specials
						switch m.Type {
						// Do lightning bounce to player
						case native.MobTypeJellyfish:
							{
								if shouldStop {
									now := time.Now()

									if now.Sub(m.JellyfishLastBounce) >= jellyfishLightningShootMS*time.Millisecond {
										wp.MobDoLightningBounce(m, v)

										m.JellyfishLastBounce = now
									}
								}
							}
						}
					}
				}
			} else {
				m.TargetEntity = nil
			}
		}

	case native.ChaoticBehavior:
		{
			m.Angle = math.Mod(m.Angle+generateValleyDistribution(-25, 25), 255)

			m.Magnitude = m.Speed() * 255 * (1. + (15.-1.)*math.Pow(rand.Float64(), 2))
		}

	case native.PassiveBehavior:
		{
			m.Magnitude = 0
		}

	case native.NeutralBehavior:
		{
			if m.LastAttackedEntity != nil {
				dx := m.LastAttackedEntity.GetX() - m.X
				dy := m.LastAttackedEntity.GetY() - m.Y

				m.Angle = TurnAngleToTarget(
					m.Angle,
					dx,
					dy,
				)

				m.Magnitude = m.Speed() * 255

				m.TargetEntity = m.LastAttackedEntity
			} else {
				m.TargetEntity = nil
			}
		}
	}
}

func generateValleyDistribution(min, max float64) float64 {
	for {
		x := min + rand.Float64()*(max-min)

		probability := math.Abs(x) / max

		if rand.Float64() < probability {
			return x
		}
	}
}
